plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -15,89 +15,165 @@
+    public IcebergSplitSource(
+            SchemaTableName schemaTableName,
+            TableScan tableScan,
+            int maxSplitsPerSecond,
+            int maxOutstandingSplits,
+            Executor executor)
     {
-        this.schemaTableName = requireNonNull(schemaTableName, "schemaTableName is null");
-        this.combinedScanIterable = requireNonNull(combinedScanIterable, "combinedScanIterable is null");
-
-        this.fileScanIterator = Streams.stream(combinedScanIterable)
-                .map(CombinedScanTask::files)
-                .flatMap(Collection::stream)
-                .iterator();
+        this.executor = requireNonNull(executor, "executor is null");
+        this.splitQueue = new ThrottledAsyncQueue<>(maxSplitsPerSecond, maxOutstandingSplits, executor);
+        this.loaderTask = new IcebergSplitLoader(
+                schemaTableName,
+                tableScan,
+                splitQueue);
+        addExceptionCallback(ResumableTasks.submit(executor, loaderTask), this::onLoaderFailure);
     }
 
     @Override
     public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
     {
-        // TODO: move this to a background thread
-        List<ConnectorSplit> splits = new ArrayList<>();
-        Iterator<FileScanTask> iterator = limit(fileScanIterator, maxSize);
-        while (iterator.hasNext()) {
-            FileScanTask task = iterator.next();
-            if (!task.deletes().isEmpty()) {
-                throw new TrinoException(NOT_SUPPORTED, "Iceberg tables with delete files are not supported: " + schemaTableName);
-            }
-            splits.add(toIcebergSplit(task));
+        checkLoaderException();
+        if (isFinished()) {
+            return CompletableFuture.completedFuture(new ConnectorSplitBatch(ImmutableList.of(), true));
         }
-        return completedFuture(new ConnectorSplitBatch(splits, isFinished()));
+        return toCompletableFuture(Futures.transform(splitQueue.getBatchAsync(maxSize), splits -> new ConnectorSplitBatch(splits, false), executor));
     }
 
     @Override
     public boolean isFinished()
     {
-        return !fileScanIterator.hasNext();
+        checkLoaderException();
+        return splitQueue.isFinished();
     }
 
     @Override
     public void close()
     {
-        try {
-            combinedScanIterable.close();
-        }
-        catch (IOException e) {
-            throw new UncheckedIOException(e);
+        checkLoaderException();
+        loaderTask.close();
+    }
+
+    private void onLoaderFailure(Throwable exception)
+    {
+        loaderException = exception;
+        loaderTask.close();
+        splitQueue.finish();
+    }
+
+    private void checkLoaderException()
+    {
+        if (loaderException != null) {
+            if (loaderException instanceof TrinoException) {
+                throw (TrinoException) loaderException;
```

Comment:
don't rethrow an exception created in different thread, the stackrace will make debugging 'funny'

wrap unconditionally